### DISCUSSION TITLE
Graphics Tracker

### DISCUSSION THREAD

**Discussion by NHodgesVFX** [bevyengine/bevy#4630] [General]
I looked through the issues and didn't see any open issues that would give an large overview of needed missing 3D graphics features in Bevy. So, I though I would create this tracker that could remain open and have items ticked of as they are added. Some features have multiple implementations that are possible, although it would be nice to have them all, I think it would be good to start with the easiest to implement in most cases. I am also sure I am forgetting some needed features as well, feel free to add them in the comments or if you have permission add them directly to this post. If this type of issue isnt wanted please feel free to close it :)

## Progress

**Rendering Methods**

- [x] Forward Clustered #3153
- [x] Deferred #9258 

**Reflections**

- [ ] Initial Screen Space Reflections (SSR) in deferred renderer #12959
- [ ] Screen Space Reflections with roughness
- [ ] Screen Space Reflections Optimizations (Half-res tracing, etc...)
- [ ] Forward Renderer Screen Space Reflections
- [ ] Planar Reflections
- [x] (HDRI) Environment Reflections, See #4581
- [ ] HDRI Pre-Filtering Built into Bevy #9414
- [x] Reflection Probes #11366
- [ ] Reflection Probe Capturing
- [ ] Raytraced Reflections

**Materials**

- [x] Support Refraction #8015
- [x] Support [KHR_materials_volume](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_materials_volume/README.md)
- [x] Support KHR_materials_transmission #8015
- [x] Support [KHR_materials_ior](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_ior) #8015
- [ ] SubSurface Scattering (SSS)
- [x] Alpha to Coverage #12970
- [x] Parallax Occlusion Mapping With Shadowing (POM) #5928
- [ ] Tessellation and Displacement
- [x] Vertex Colors #4528
- [ ] Decals #2401

**Import**

- [x] Initial Auto Mip Map Generation
- [ ] Auto Mip Map Generation
- [ ] Auto UV Generation
- [x] Auto Normal Generation #3987 #11654
- [x] Auto Tangent Generation #3872
- [ ] Acsii Embedded GTLF Support
- [x] Texture Compression

**Occlusion**

- [x] Screen Space Ambient Occlusion (SSAO) #3632
- [x] Ground Truth Ambient Occlusion (GTAO) #7402
- [ ] RayTraced Ambient Occlusion (RTAO)
- [ ] Bent Normals 
- [x]  Approximate indirect specular occlusion #11152 

**Global Illumination**

- [ ] Screen Space Global Illumination (SSGI)
- [ ] Voxel Global Illumination (VXGI) 
- [ ] SDF Global Illumination (SDFGI)
- [x] Lightmaps  #10231
- [ ] Directional Lightmaps
- [ ] Lightmap Baking
- [x] Light Probes #11366 #10268
- [x] Irradiance Volumes #10268
- [ ] Light Probe "Baking" 
- [ ] RayTraced Global Illumination (RTGI)

**Lights**

- [x] Spot Lights #4715
- [ ] Rect Area Light
- [ ] Tube or Cylindrical Area Light
- [x] Image Based Lighting, See #4581

**Shadows**

- [x] Cascaded Shadow Mapping #3629
- [ ]  Sample-Distribution Shadow Maps #4253
- [x] Percentage Closer Filtering (PCF)  #3628 #8006
- [x] PCF for point lights #12910
- [ ] Percentage Closer Soft Shadows (PCSS) #3631 and #13497
- [ ] RayTraced Shadows
- [ ] Screen Space Shadows (Contact Shadows)

**Effects**

- [x] Fast approximate anti-aliasing (FXAA) #6393
- [x] Subpixel Morphological Antialiasing (SMAA) #9819
- [x] Temporal Anti-Aliasing (TAA) #3663
- [x] Bloom #6397
- [x] Tone Mapping #3425
- [x] Depth Of Field (https://github.com/bevyengine/bevy/pull/13009)
- [ ] Color Correction
- [x] Color Grading
- [x] Volumetric Lighting (https://github.com/bevyengine/bevy/pull/13057)
- [x] Depth Fog And Atmospheric #6412
- [x] Auto Exposure / Eye Adaptation #12792
- [ ] Upscaling - DLSS, XeSS, FSR 1, 2, & 3
- [x] Dithering #5264
- [x] Robust Constrast Adaptive Sharpening (RCAS)
- [ ] Atmosphere / Sky rendering

**Animation**

- [x] GPU Skinning
- [x] Skeletal
- [x] Morph Targets 
- [ ] Animation #4026

**Optimizations**

- [ ] Occlusion Culling
- [ ] Hardware Instancing #89
- [ ] Level Of Detail (LOD) #6868 
- [x] Hierarchical Level of Detail (HLOD) https://github.com/bevyengine/bevy/pull/12916
- [ ] Asset Streaming #8530
- [ ] Compute Shader Skinning #7502
- [ ] GPU-driven Rendering
- [ ] Virtual Textures
- [ ] Shadow Atlases
- [ ] Virtual Shadow Maps
- [ ] Caching Shadow Maps
- [x] Automatic batching/instancing of draw commands #9685

**2 & 2.5D**
- [ ] Lighting & Shading
- [ ] Global Illumination

## Additional context

**Refraction**

- https://www.khronos.org/news/press/new-gltf-extensions-raise-the-bar-on-3d-asset-visual-realism

**Comment by alice-i-cecile** [comment]
This is an *excellent* list. I'll defer to @superdump, our rendering lead, for the best format for this.

You may also be interested in the [rendering project board](https://github.com/orgs/bevyengine/projects/3), which is a more operational view of the work that's being done.

**Comment by NHodgesVFX** [comment]
Thanks for the board its really interesting, I had found some of those issues just by searching but their are quite a few I didn't see before.

**Comment by alice-i-cecile** [comment]
If you have the expertise @NHodgesVFX, PR reviews from the community are incredibly valuable to help contributors learn and enable us to merge in features faster. This is especially true in rendering, where the team's expertise can often run a bit thin.

**Comment by NHodgesVFX** [comment]
Thanks, However I don't really have the expertise, I'm pretty new to rust and graphics programming. But maybe I will have the expertise sometime in the future :)

**Comment by james7132** [comment]
We're already skinning on the GPU.  Unless the suggestion is to use bindless rendering to skin meshes, we're already doing this.

Animation blending and morph target work is already being tracked by #4026.

**Comment by NHodgesVFX** [comment]
I updated the list to reflect that, didn't realize skinning was already done on the GPU.

**Comment by cart** [comment]
Great list! I do think this feels more like Github Discussion territory rather than formal tracking issue territory. "3D graphics features" is a nebulous, large, and ever-changing area that we will never really be able to drive down to zero and close out. Any objections to me converting this to a discussion? You can still use this to track progress, but issues should be "scoped, well defined, and with clear paths to resolution".

**Comment by alice-i-cecile** [comment]
Yeah, I like this as a discussion too :)

**Comment by superdump** [comment]
A bunch of these have open issues and even PRs. I think they are all on the rendering project board. Could you add links to them all to the original post? I can’t edit it so I can’t maintain it either, whereas the project board is something I do keep on top of.

**Reply by NHodgesVFX** [reply]
Sure Ill add the links I see, no worries Ill try to keep it up to date as best I can.

**Comment by kirawi** [comment]
Are any of these good first issues?

**Reply by alice-i-cecile** [reply]
Unlikely, although if you have rendering expertise they'd probably be manageable :) #rendering-dev on [Discord](https://discord.com/invite/bevy) is quite helpful and does a good job teaching new contributors the ropes; I expect they'll have suggestions.

**Comment by superdump** [comment]
There's now a PR for spotlights: https://github.com/bevyengine/bevy/pull/4715

**Comment by JMS55** [comment]
For tonemapping, auto-exposure by calculating the luminance of each frame would be neat.

**Reply by NHodgesVFX** [reply]
Good call, added.

**Comment by JMS55** [comment]
To upscalers/external sdks, we should add FSR, XeSS, and DLSS (if the license allows us).

**Reply by minecrawler** [reply]
- [FidelityFX-FSR](https://github.com/GPUOpen-Effects/FidelityFX-FSR) is available as source using [MIT](https://github.com/GPUOpen-Effects/FidelityFX-FSR/blob/master/license.txt)
- [FidelityFX-FSR2](https://github.com/GPUOpen-Effects/FidelityFX-FSR2) is available as source using [MIT](https://github.com/GPUOpen-Effects/FidelityFX-FSR2/blob/master/LICENSE.txt)
- [XeSS](https://github.com/intel/xess) is available as SDK using [Intel Simplified Software License](https://github.com/intel/xess/blob/main/licenses/LICENSE.pdf), which looks like it should be compatible
- [DLSS 2](https://github.com/NVIDIAGameWorks/Streamline) is available as SDK using [MIT](https://github.com/NVIDIAGameWorks/Streamline/blob/main/license.txt)
- [DLSS 3](https://developer.nvidia.com/rtx/dlss) is not available, yet, but it may become part of the Streamline SDK under MIT

**Reply by minecrawler** [reply]
To be honest, I would love to see a basic implementation, which works everywhere, in Bevy (for example custom or FSR), and have the others (FSR1/2, XeSS, DLSS2/3) available as plugins.

**Reply by NHodgesVFX** [reply]
Great Suggestions, I added them to the list.

**Comment by JMS55** [comment]
FXAA has been merged in https://github.com/bevyengine/bevy/pull/6393, and Bloom has been merged in https://github.com/bevyengine/bevy/pull/6397 :tada:.

**Reply by NHodgesVFX** [reply]
Thanks for letting me know :) Updated the post

**Comment by matt328** [comment]
I'm curious about a plan for tessellation and displacement, with wgpu only supporting the basic pipeline stages, would things that use tessellation control and eval / hull and domain shaders need a different rendering library altogether?  Wgpu seems like a great abstraction over lower level rendering APIs like OpenGL, Vulkan and DirectX, but I think the lack of support for tessellation and/or mesh shaders maybe become somewhat limiting.  I'm admittedly not familiar with bevy's roadmap or future plans, so take all this with a grain of salt.

**Reply by NHodgesVFX** [reply]
You can do tessellation with a compute shader, which WGPU supports.

https://onrendering.com/data/papers/isubd/isubd.pdf

Also WGPU might add a extension for tesselation
https://github.com/gfx-rs/wgpu/issues/222

**Comment by JMS55** [comment]
Cascaded shadows maps are now complete. We should also add Sample-Distribution Shadow Maps https://github.com/bevyengine/bevy/issues/4253 to the list.

**Reply by NHodgesVFX** [reply]
Thanks, I updated the tracker.

**Comment by wdhwg001** [comment]
Perhaps we could have a section mentioning that we have Clustered Forward Rendering (#3153 ) as the only one renderer.

And some other updates:

GPU instancing is tracked in #89 

Animations are tracked in #4026 

Dithering is finished in #5264 

Distance and Atmospheric Fog is finished in #6412 

Tonemapping is done and made into a separated mod in #3425 

Parallax Mapping has a draft in #5928

**Reply by NHodgesVFX** [reply]
Thanks, Updated!

**Comment by JMS55** [comment]
Deferred: #9258
HDRI Pre-Filtering Built into Bevy: #9414 
Ground Truth Ambient Occlusion (GTAO): Covered by the original SSAO PR

**Reply by NHodgesVFX** [reply]
Thanks Updated!

**Comment by JMS55** [comment]
For shadows, we should add screen space shadows, sometimes called contact shadows.

**Reply by NHodgesVFX** [reply]
Thanks for the suggestion, added!

**Comment by JMS55** [comment]
Under optimization, please add :

* Compute shader skinning https://github.com/bevyengine/bevy/issues/7502
* A general "GPU-driven rendering" task (no dedicated tracking issue, but there's some thoughts in https://github.com/bevyengine/bevy/issues/89#issuecomment-1379249074. Probably just don't link an issue for now).
* Asset streaming (meshes, materials, and textures), I suppose linking to https://github.com/bevyengine/bevy/issues/8530, but rendering needs special considerations tied into the culling systems.
* Shadow atlases/virtual shadow maps, and/or caching shadowmaps (these are all kind of linked)
* Level of detail https://github.com/bevyengine/bevy/issues/6868

Under import, please add: Texture compression - now that Assets V2 has been merged, we have the ability to have asset metadata and preprocess assets. We should take material textures and preprocess them into astc/etc2/bcn/etc.

Under materials, please add decals https://github.com/bevyengine/bevy/issues/2401

Under effects, please add "Atmosphere/sky rendering"

**Reply by PlexusDuMenton** [reply]
You could also add Virtual Textures since you are talking about Virtual Shadow map

**Reply by NHodgesVFX** [reply]
Thanks everyone for the suggestions they have been added.

**Comment by samoylovfp** [comment]
Could Mipmap generation and texture compression be ticked off?

https://bevyengine.org/news/bevy-0-12/#compressedimagesaver
>Bevy 0.12 ships with a barebones [CompressedImageSaver](https://docs.rs/bevy/0.12.0/bevy/render/texture/struct.CompressedImageSaver.html) that writes images to [Basis Universal](https://github.com/BinomialLLC/basis_universal) (a GPU-friendly image interchange format) and generates [mipmaps](https://en.wikipedia.org/wiki/Mipmap).

**Reply by alice-i-cecile** [reply]
I wouldn't call that fully complete. I'd want a more automated and battle-tested mip-map generation process first.

**Reply by NHodgesVFX** [reply]
Thanks for the suggestions. I checked off texture compression and added a new task, which is checked off, for the initial implementation of MIP maps.

**Reply by PlexusDuMenton** [reply]
In addition, Deferred and Support Refraction should be ticked, (and likely "Support KHR_materials_transmission" and "Support KHR_materials_ior" but I didn't checked properly if they are finished)

Edit: PCF shadow for Directional & spotlight is also done (missing point lights)

**Reply by NHodgesVFX** [reply]
Thanks, Updated!

**Comment by mrgzi** [comment]
I don't know much about 3D graphic concept. Maybe this question is not related with this discussion but I am currious that can Bevy compete with Unreal Engine graphics when this list completed totally?

**Reply by NHodgesVFX** [reply]
Simple Answer, Probably Not if all your considering is graphical fidelity and realism. Once this list is completed as written now, Unreal Engine will likely have new graphics features which would need to be added. However, once this list is completed Bevy would be in a far better position to support all styles of games.  Its completely possible to make a game with bevy but at the moment you are a bit limited to certain styles. Take a look at the Godot Engine which has most of the features on this list, its possible to make somewhat real looking stuff and stylized stuff already but as far as pure graphical quality Unreal still wins. I would argue Unreal is graphics wise superior to Unity as well.

Also, I would like to  mention I may be biased do to where I work

**Reply by mrgzi** [reply]
Thanks for your answer. Also, why there is no 2D graphics tracker?

**Reply by NHodgesVFX** [reply]
Advanced rendering features are less needed in 2D then in 3D. With 2D, most games rely on art and tricks, or custom solutions rather then advanced rendering. That being said 2D lighting and shading would be nice to see and of course there are other features important to 2D like batching, which was implemented recently. However, this is a good point I might rename this discussion to Graphics Tracker and add a 2D section.

For the most part in Bevy its possible to make pretty much any 2D game.

**Reply by mrgzi** [reply]
It would be great if you can also add 2D section. Maybe also 2.5D section?

**Reply by NHodgesVFX** [reply]
Thanks for the suggestion I added a combined 2 and 2.5D section to the tracker. Also, sorry it took a while to add. Also, renamed the tracker to Graphics Tracker.

**Reply by wdhwg001** [reply]
@NHodgesVFX Thanks for adding the 2D section! I was wondering if we could have 2D skeletal animation on the tracker list.

Also, the 2D section probably needs to have a Pixel Perfect feature, which is not built in but included in an example: https://github.com/bevyengine/bevy/pull/8112

**Comment by TirushOne** [comment]
@NHodgesVFX non directional lightmaps recentley got merged so you should probobley tick that one off

**Reply by NHodgesVFX** [reply]
Thanks for letting me know, Sorry for the wait, the tracker has been updated.

**Comment by pcwalton** [comment]
Lightmaps landed in #10231. Reflection probes and light probes landed in #11366. Voxel global illumination (VXGI) is #10268 and is on the milestone for 0.13.

**Reply by NHodgesVFX** [reply]
Thanks, Updated!

**Comment by pcwalton** [comment]
Auto exposure is submitted as a PR here: https://github.com/bevyengine/bevy/pull/12792

PCF for point lights landed in: https://github.com/bevyengine/bevy/commit/d59b1e71ef6717a530e4c117c7e58495100eb480

LOD support equivalent to Unity's is available as a PR: https://github.com/bevyengine/bevy/pull/12916
There's still more that could be done, however.

**Reply by NHodgesVFX** [reply]
Thanks, Updated!

**Comment by pcwalton** [comment]
Alpha to coverage is https://github.com/bevyengine/bevy/pull/12970

Basic screen space reflections are https://github.com/bevyengine/bevy/pull/12959

**Reply by NHodgesVFX** [reply]
Thanks, Updated!

**Reply by pcwalton** [reply]
https://github.com/bevyengine/bevy/commit/1141e731ff19b74a05a6290f71cdb2a0a3b74b8d -- Alpha to Coverage -- is landed now

**Comment by pcwalton** [comment]
HLOD has landed https://github.com/bevyengine/bevy/pull/12916

Color grading has landed (and color correction too? not sure what you consider the difference to be) https://github.com/bevyengine/bevy/pull/13121

Auto exposure has landed https://github.com/bevyengine/bevy/pull/12792
